### PR TITLE
ISLANDORA-2140: Prevent garbled MODS when going back and forth in ingest forms

### PR DESCRIPTION
**Pull request by d-r-p**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-2140

# What does this Pull Request do?

Address [ISLANDORA-2140](https://jira.duraspace.org/browse/ISLANDORA-2140), i.e. prevent garbled MODS when going back and forth in ingest forms.

In multi-paged ingest forms, when the user returns to the metadata form and re-submits, the original code swallows certain XML sub-elements inside the resulting MODS. More precisely, whenever an XML fragment of the form `<element><sub-element>%value%</sub-element></element>` is created, going back to and re-submitting the metadata form results in a MODS containing only `<element>the inserted value</element>`. This breaks the MODS standard, along with subsequent logic such as CSL-driven metadata views.

The proposed solution fixes this behaviour.

# What's new?

When creating an XML fragment:
* adjust the DOMNode to register in the [NodeRegistry](https://github.com/Islandora/islandora_xml_forms/blob/7.x/api/NodeRegistry.inc) of the handled [XMLDocument](https://github.com/Islandora/islandora_xml_forms/blob/7.x/api/XMLDocument.inc)
* update the FormElement's value with the [Read](https://github.com/Islandora/islandora_xml_forms/blob/7.x/api/Read.inc)-value of the DOMNode just created inside the DOMDocument

If, during ingest, the user goes back to the metadata form, those adjustments allow the [Update action](https://github.com/Islandora/islandora_xml_forms/blob/7.x/api/Update.inc) triggered upon re-submission to work correctly.

# How should this be tested?

* Convince yourself of the issue by following the ten steps described in the [JIRA ticket](https://jira.duraspace.org/browse/ISLANDORA-2140)
* Apply the diff and assess that the problem has disappeared
* Test other ingest forms (content models)

Of course, do not test directly on a production environment!

_Note_: The modifications change the [NodeRegistry](https://github.com/Islandora/islandora_xml_forms/blob/7.x/api/NodeRegistry.inc) built during DOMNode creation ([Create action](https://github.com/Islandora/islandora_xml_forms/blob/7.x/api/Create.inc)) in a substantial manner. Therefore, **regression testing** might be advisable.

# Additional Notes:
* Does this change require documentation to be updated? **no**
* Does this change add any new dependencies? **no**
* Does this change require any other modifications to be made to the repository (ie. Regeneration activity, etc.)? **no**
* Could this change impact execution of existing code? **possibly**

# Interested parties
@DiegoPino, @Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by DiegoPino** [comment]
@d-r-p cool, code is nice!! still, need to understand [what is happening there and your suggested solution](https://jira.duraspace.org/browse/ISLANDORA-2140?focusedCommentId=58316&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-58316). Will ask you a few questions once I get to test this. 
For now, and forgive my ignorance on the approach, i'm not sure if `self` as xpath concept is refering actually to "what a field created" or to the "field itself". It is tricky since this form fields can create single nodes of course but also complex trees. Which means if the field (or a human through the field really) is adding a complex, deep nested snipped of XML, `self`, still refers to the "parents node value"... but since its a node, `self::node` would be more true to XML xpath reality? Also the reason (not only because i have no life) i did not keep doing this and got a bit lost. Let me test this and come back to you this week. We have 7.x committers call and i can get some extra opinions. Committers call is on Thursday at 2PM EST in case you would like to explain this to others in digital-person. Thanks a lot.

**Comment by d-r-p** [comment]
@DiegoPino If I understand your comment/question correctly, the issue is precisely there: during Create, self::node() points to a parent node of the (sub-)element we want to create, whereas during Update, self::node() points to the node pointed at by the Read action (which should be the (sub-)element the value of which we wish to access through the form). This discrepancy is responsible for the problematic behaviour, since the contexts are not re-built when going back to the metadata form, but taken from cache. As far as I understand, the contexts are basically stored in the NodeRegistry of the handled XMLDocument so that they persist across requests. My update gets rid of the aforementioned discrepancy.
I will be happy to answer to your questions (here, or via PM). Let me know...

**Comment by DiegoPino** [comment]
@d-r-p yes, you understood my question much better than I managed to formulate it. Let me give it a spin and I will come to you. Thanks again

**Comment by DiegoPino** [comment]
@d-r-p this looks good.! I will do some more tests tonight with some deeper XML snippets and see what it means (and also why we never saw this failing..) but it does what it says, thanks a lot. In the meantime, can you please submit an[ iCLA](https://github.com/Islandora/islandora/wiki/Contributor-License-Agreements) to be able to merge this once it is tested and send it to community@islandora.ca? @rosiel @dannylamb is that the email the CLA should go? Thanks

**Comment by dannylamb** [comment]
@d-r-p's CLA is on file and good to go.  If you feel this PR is :+1: , then ship it.  :shipit:

**Comment by jordandukart** [comment]
This looked fine to me as well.

**Review by DiegoPino — APPROVED** [review]
This actually fixes a complicated to trigger Bug. It is also complicated to explain the how and the why so extra kudos @d-r-p for this awesome contribution. Checked!

**Pull request merged**
